staging/src/k8s.io/client-go/rest/request_test.go

```diff
@@ -986,56 +986,22 @@ func TestRequestWatch(t *testing.T) {
 			},
 			serverReturns: []responseErr{
 				{response: &http.Response{
+					Header:     http.Header{"Content-Type": []string{"application/json"}},
 					StatusCode: http.StatusForbidden,
-					Body:       io.NopCloser(bytes.NewReader([]byte{})),
```

Comment:
the change in staging/src/k8s.io/client-go/rest/request.go looks reasonable

I am surprised that we're deleting a test case and changing another existing testcase for this... I would have expected to leave existing tests ~as-is and *add* a new test case (or *add* checks to the existing testcases) expecting the more verbose errors

Reply:
The deleted test case "server returns forbidden" below is same as the updated test case "server returns forbidden" except the latter had an extra `Expect` of `Event` which is wrong and test passed only because it was not checked when `Err` is `true`.

To make it return the correct error like running in production, the response header should contain `Content-Type` header and a json encoded `Status` body. That's why the existing test case's inputs need update. If we want to leave existing tests as is, I think the duplicate test could be kept (but needs to be renamed to avoid confusion) to validate when server returns an empty body, and another test's inputs could be updated to test normal server response.

Reply:
> To make it return the correct error like running in production, the response header should contain Content-Type header and a json encoded Status body.

I'm not sure I buy this... we should make sure the new code handles weird empty error responses in a sensible way.

1. Keep the existing test case returning `io.NopCloser(bytes.NewReader([]byte{})),`, and adjust the test's expectation to show what the resulting error will be with the change in this PR
2. Add a *new* test case with a json-containing error response with content-type, showing what the resulting error will be for a response like that

Reply:
> we should make sure the new code handles weird empty error responses in a sensible way.
> 
> 1. Keep the existing test case returning `io.NopCloser(bytes.NewReader([]byte{})),`, and adjust the test's expectation to show what the resulting error will be with the change in this PR
> 2. Add a _new_ test case with a json-containing error response with content-type, showing what the resulting error will be for a response like that

I agree with you and I think this is what exactly the patch is doing. It appears like it modifies the existing test because there were two same tests. The following two code blocks are from the master branch:
https://github.com/kubernetes/kubernetes/blob/de6db3ffb0680f2022f8795038eb2ca5b36c5bd6/staging/src/k8s.io/client-go/rest/request_test.go#L980-L992
https://github.com/kubernetes/kubernetes/blob/de6db3ffb0680f2022f8795038eb2ca5b36c5bd6/staging/src/k8s.io/client-go/rest/request_test.go#L1023-L1035

So I change one of them to test with a json-containing error response with content-type and keep another one to test with empty error response. The following two code blocks are from the patch:
https://github.com/kubernetes/kubernetes/blob/596c5696c64023808af164284263647d795b0ac2/staging/src/k8s.io/client-go/rest/request_test.go#L980-L998
https://github.com/kubernetes/kubernetes/blob/596c5696c64023808af164284263647d795b0ac2/staging/src/k8s.io/client-go/rest/request_test.go#L1009-L1021

Reply:
gotcha, sorry for the confusion